File: src/astprinter.cpp
#include "astprinter.hpp"
#include "expr.hpp"

#include <memory>
#include <string>

std::string AstPrinter::print(std::shared_ptr<Expr<std::string>> expr) {
    return expr->accept(std::make_shared<AstPrinter>(*this));
}

std::string AstPrinter::visitLiteralExpr(std::shared_ptr<Literal<std::string>> expr) {
    return expr->value.__str__();
}

std::string AstPrinter::visitGroupingExpr(std::shared_ptr<Grouping<std::string>> expr) {
    return "( group " + expr->accept(std::make_shared<AstPrinter>(*this)) + " )";
}

std::string AstPrinter::visitUnaryExpr(std::shared_ptr<Unary<std::string>> expr) {
    return "(" + expr->operation.lexme + " " + expr->right->accept(std::make_shared<AstPrinter>(*this)) + ")";
}

std::string AstPrinter::visitBinaryExpr(std::shared_ptr<Binary<std::string>> expr) {
    return "(" + expr->left->accept(std::make_shared<AstPrinter>(*this)) + " " +
                 expr->operation.lexme + " " +
                 expr->right->accept(std::make_shared<AstPrinter>(*this)) + ")";
}

File: src/ast_printer.py
from .expr_visitor import Visitor
from .expr import Expr, Binary, Grouping, Literal, Unary, Ternary, Postfix
from .token import Token, TokenType


class AstPrinter(Visitor):
    def print(self, expr: Expr) -> str:
        return expr.accept(self)

    def visitBinaryExpr(self, expr: Binary) -> str:
        return f"({expr.left.accept(self)} {expr.operator.lexme} {expr.right.accept(self)})"

    def visitGroupingExpr(self, expr: Grouping) -> str:
        return f"({expr.expression.accept(self)})"

    def visitUnaryExpr(self, expr: Unary) -> str:
        return f"({expr.operator.lexme} {expr.expression.accept(self)})"

    def visitLiteralExpr(self, expr: Literal) -> str:
        return f"{expr.value}"

    def visitTernaryExpr(self, expr: Ternary) -> str:
        return f"({expr.one.accept(self)} {expr.op1.lexme} {expr.two.accept(self)} {expr.op2.lexme} {expr.three.accept(self)})"

    def visitPostfixExpr(self, expr: Postfix):
        return f"({expr.expression.accept(self)} {expr.operator.lexme})"

def main():
    expr = Binary(
        Unary(
            Token(TokenType.MINUS, '-', None, 1),
            Literal(123)
        ),
        Token(TokenType.STAR, '*', None, 1),
        Grouping(
            Literal(45.67)
        )
    )

    print(AstPrinter().print(expr))
